topology/src/main/java/io/camunda/zeebe/topology/PersistedClusterTopology.java

```diff
@@ -33,20 +73,89 @@ void update(final ClusterTopology clusterTopology) throws IOException {
     if (this.clusterTopology.equals(clusterTopology)) {
       return;
     }
+    writeToFile(clusterTopology);
+    this.clusterTopology = clusterTopology;
+  }
+
+  public boolean isUninitialized() {
+    return clusterTopology.isUninitialized();
+  }
+
+  private static ClusterTopology readFromFile(
+      final Path topologyFile, final ClusterTopologySerializer serializer) throws IOException {
+    if (!Files.exists(topologyFile)) {
+      return ClusterTopology.uninitialized();
+    }
+
+    final var content = Files.readAllBytes(topologyFile);
+    if (content.length < HEADER_LENGTH) {
+      throw new MissingHeader(topologyFile, content.length);
+    }
+    final var header = ByteBuffer.wrap(content, 0, HEADER_LENGTH).order(ByteOrder.LITTLE_ENDIAN);
+    final var version = header.get();
+    final var expectedChecksum = header.getLong();
+
+    if (version != VERSION) {
+      throw new UnexpectedVersion(topologyFile, version);
+    }
+
+    final var body = Arrays.copyOfRange(content, HEADER_LENGTH, content.length);
```

Comment:
:question: Why do we need to copy the range? It would be better if we could decode already from a given offset/length, since the buffer is "fresh" so to speak, and only for this case. But as it's not in the hot path, it's probably fine :shrug:

Reply:
Ah yeah, I went back and forth on operation over byte arrays vs. passing around a byte buffer.
Passing around the offset/length seems like a good compromise :+1: